### PR TITLE
[Feature] Dynamic type handling for Raw Queries.

### LINKED ISSUE #1591: [Feature] Support of dynamic type in raw queries.
Original Stackoverflow issue:
https://stackoverflow.com/questions/54530251/linq2db-and-query-for-dynamic

```cs
public IDictionary<string, int> GetParentelaMapping()
    {
        using (var conn = dataContextFactory.Create())
        {
            var result = conn.Query<dynamic>("SELECT ID_GRADO_PARENTELA,GRADO_PARENTELA FROM GRADO_PARENTELA")
                .ToDictionary(
                    row => (string)row.GRADO_PARENTELA,
                    row => (int)row.ID_GRADO_PARENTELA, StringComparer.OrdinalIgnoreCase);

            return result;
        }
    }
```

### PR DESCRIPTION
**Pull request by sdanyliv**
Fixes #1591

### CODE REVIEW

File: Source/LinqToDB/Data/CommandInfo.cs [hunk 1]
```diff
@@ -1662,6 +1690,61 @@ select new
 			return lex.Compile();
 		}
 
+
+		static ConstructorInfo _expandoObjectConstructor = MemberHelper.ConstructorOf(() => new ExpandoObject());
+		static MethodInfo      _expandoAddMethodInfo     = MemberHelper.MethodOf(() => ((IDictionary<string, object>)null!).Add("", ""));
```
Comment (Shane32) [inline Source/LinqToDB/Data/CommandInfo.cs:1695]:
```suggestion
		static readonly ConstructorInfo _expandoObjectConstructor = MemberHelper.ConstructorOf(() => new ExpandoObject());
		static readonly MethodInfo      _expandoAddMethodInfo     = MemberHelper.MethodOf(() => ((IDictionary<string, object>)null!).Add("", ""));
```
readonly?

**Review by Shane32 — APPROVED** [review]
Looks great!

File: Tests/Linq/Linq/DynamicResultTests.cs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+			public static RawDynamicData[] Seed()
+			{
+				return Enumerable.Range(1, 20)
+					.Select(i => new RawDynamicData { AId = i, BId = i * 100, AValue = i * 2, BValue = i * 100 * 2 })
+					.ToArray();
+			}
+		}
+
+		[Test]
+		public void DynamicQueryViaDynamic([IncludeDataSources(TestProvName.AllSQLite)] string context, [Values(1, 10)] int param)
```
Comment (MaceWindu) [inline Tests/Linq/Linq/DynamicResultTests.cs:30]: add sqlserver 2019 sequential access provider to tests to test CreateDynamicObjectReader

File: Source/LinqToDB/Data/CommandInfo.cs [hunk 3]
```diff
@@ -324,16 +325,21 @@ public async Task<IEnumerable<T>> QueryAsync<T>(CancellationToken cancellationTo
...
 			using (scope)
 				try
 				{
 					if (rd.Read())
 					{
-						var additionalKey = GetCommandAdditionalKey(rd);
+						var additionalKey = GetCommandAdditionalKey(rd, typeof(T));
 						var objectReader  = GetObjectReader<T>(DataConnection, rd, DataConnection.Command.CommandText,
-							additionalKey);
+							additionalKey, isSingleValue: false);
```
Comment (MaceWindu) [inline Source/LinqToDB/Data/CommandInfo.cs:342]: rename isSingleValue to isScalar?

**Review by MaceWindu — CHANGES_REQUESTED** [review]

File: Source/LinqToDB/Data/CommandInfo.cs (hunk 3 above)
Comment (sdanyliv) [reply Source/LinqToDB/Data/CommandInfo.cs:342]: Make sense.

File: Tests/Linq/Linq/DynamicResultTests.cs (hunk 2 above)
Comment (sdanyliv) [reply Tests/Linq/Linq/DynamicResultTests.cs:30]: I can do that for possible regressions. in the future. But originally algorithm read data exactly in sequential order.

**Review by viceroypenguin — APPROVED** [review]

**Review by igor-tkachev — APPROVED** [review]

**Review by MaceWindu — APPROVED** [review]

**Pull request merged**